web/docs/project/testing.md

```diff
@@ -1,3 +1,171 @@
+
+  [`@testing-library/react`](https://www.npmjs.com/package/@testing-library/react) / [`@testing-library/jest-dom`](https://www.npmjs.com/package/@testing-library/jest-dom): Testing helpers.
+
+  [`msw`](https://www.npmjs.com/package/msw): A server mocking library.
+
+  </div>
+</details>
+
+### Setting Up Tests
+
+In order for Wasp to pick up your tests, they should be placed within the `src/client` directory and use an extension that matches [these glob patterns](https://vitest.dev/config#include). Some of the file names that Wasp will pick up as tests:
+
+- `yourFile.test.ts`
+- `YourComponent.spec.jsx`
+
+Within test files, you can import your other source files as usual. For example, if you have a component `Counter.jsx`, you test it by creating a file in the same directory called `Counter.test.jsx` and import the component with `'./Counter'`.
+
+### Running Tests
+
+Running `wasp test client` will start Vitest in watch mode and recompile your Wasp project when changes are made.
+
+- If you want to see a realtime UI, pass `--ui` as an option.
+- To run the tests just once, use `wasp test client run`.
+
+All arguments after `wasp test client` are passed directly to the Vitest CLI, so check out [their documentation](https://vitest.dev/guide/cli.html) for all of the options.
+
+:::warning
+You should not run `wasp test` while `wasp start` is running. Both will try to compile your project to `.wasp/out`.
+:::
+
+### React Testing Helpers
+
+Wasp provides several functions to help you write React tests:
+
+- `mockQuery`: Takes a Wasp Query to mock and the JSON data to return.
+  - This is helpful if your Query uses `useQuery`.
+  - Behind the scenes, Wasp uses [`msw`](https://npmjs.com/package/msw) to create a server request handler that responds with the given JSON.
+  - Mocks are cleared after each test.
+- `mockApi`: Similar to `mockQuery`, but for [APIs](/docs/advanced/apis) instead.
+  - Instead of a Wasp Query, it takes a route of the shape: `mockApi({ method: HttpMethod.Get, path: '/foo/bar' }, { res: 'hello' })`.
+  - You can import `HttpMethod` like so: `import { HttpMethod } from '@wasp/types'`.
+- `renderInContext`: Takes a React component, wraps it inside a `QueryClientProvider` and `Router`, and renders it.
+
+## Testing Your Server-Side Code
+
+Coming soon!
+
+## Examples
+
+### Client Unit Tests
+
+<Tabs groupId="js-ts">
+<TabItem value="js" label="JavaScript">
+
+```js title="src/client/Todo.test.js"
+import { test, expect } from 'vitest'
+
+import { areThereAnyTasks } from './Todo'
+
+test('areThereAnyTasks', () => {
+  expect(areThereAnyTasks([])).toBe(false)
+})
+```
+
+</TabItem>
+<TabItem value="ts" label="TypeScript">
+
+```ts title="src/client/Todo.test.ts"
+import { test, expect } from 'vitest'
+
+import { areThereAnyTasks } from './Todo'
+
+test('areThereAnyTasks', () => {
+  expect(areThereAnyTasks([])).toBe(false)
+})
+```
+
+</TabItem>
+</Tabs>
+
+### React Component Tests
```

Comment:
We don't have an example for using the `mockApi` helper, maybe add as well.

Reply:
Check out the new docs merged in the parent branch, they should be better.

I see APIs an escape hatch which enables defining Express routes. It just "raw" backend endpoints.

They are used from the client by using `axios` via our `api` wrapper. So, `mockApi` should mock some JSON e.g. list of TODOs. And then the component is expected to render a list of `li`s.